src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/Functions.java

```diff
@@ -135,259 +148,349 @@
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.subsequence2;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.subsequence3;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.substring2;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.substring3;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.substring_after;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.substring_before;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.sum1;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.sum2;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.tail;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.tan;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.text_file1;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.text_file2;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.time;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.timezone_from_date;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.timezone_from_dateTime;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.timezone_from_time;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.tokenize1;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.tokenize2;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.values;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.yearMonthDuration;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.year_from_date;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.year_from_dateTime;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.years_from_duration;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.zero_or_one;
 
 public class Functions {
-    private static HashMap<FunctionIdentifier, Class<? extends RuntimeIterator>> builtInFunctions;
+    private static final HashMap<FunctionIdentifier, BuiltinFunction> builtInFunctions;
     private static HashMap<FunctionIdentifier, FunctionItem> userDefinedFunctions;
 
+    private static final Map<String, ItemType> itemTypes;
+
+    static {
+        itemTypes = new HashMap<>();
+        itemTypes.put("item", new ItemType(ItemTypes.Item));
```

Comment:
Caching types and indexing them in this way is a very good idea both in terms of performance and productivity.